cloud/blockstore/libs/endpoints/endpoint_manager.cpp

```diff
@@ -837,59 +845,130 @@ NProto::TError TEndpointManager::AlterEndpoint(
+        oldReq.SetVolumeMountMode(newReq.GetVolumeMountMode());
+        oldReq.SetMountSeqNumber(newReq.GetMountSeqNumber());
+
+        if (!CompareRequests(newReq, oldReq)) {
+            return MakeError(E_INVALID_STATE, TStringBuilder()
+                << "endpoint " << socketPath.Quote()
+                << " has already been started with other args");
+        }
+
+        changed = true;
+        error = AlterEndpoint(ctx, newReq);
+        if (HasError(error)) {
+            return error;
+        }
     }
 
-    oldReq.SetVolumeAccessMode(newReq.GetVolumeAccessMode());
-    oldReq.SetVolumeMountMode(newReq.GetVolumeMountMode());
-    oldReq.SetMountSeqNumber(newReq.GetMountSeqNumber());
+    if (!NFs::Exists(socketPath)) {
+        changed = true;
+        error = RecreateSocket(ctx, newReq);
+        if (HasError(error)) {
+            return error;
+        }
+    }
 
-    if (!CompareRequests(newReq, oldReq)) {
-        return MakeError(E_INVALID_STATE, TStringBuilder()
+    return changed
+        ? error
+        : MakeError(S_ALREADY, TStringBuilder()
             << "endpoint " << socketPath.Quote()
-            << " has already been started with other args");
-    }
+            << " has already been started");
+}
+
+NProto::TError TEndpointManager::AlterEndpoint(
+    TCallContextPtr ctx,
+    const NProto::TStartEndpointRequest& request)
+{
+    STORAGE_INFO("Alter endpoint: " << request);
 
     auto future = SessionManager->AlterSession(
         ctx,
-        socketPath,
-        newReq.GetVolumeAccessMode(),
-        newReq.GetVolumeMountMode(),
-        newReq.GetMountSeqNumber(),
-        newReq.GetHeaders());
+        request.GetUnixSocketPath(),
+        request.GetVolumeAccessMode(),
+        request.GetVolumeMountMode(),
+        request.GetMountSeqNumber(),
+        request.GetHeaders());
 
     if (auto error = Executor->WaitFor(future); HasError(error)) {
         return error;
     }
 
-    auto [sessionInfo, error] = Executor->WaitFor(SessionManager->GetSession(
+    auto sessionFuture = SessionManager->GetSession(
         ctx,
-        socketPath,
-        newReq.GetHeaders()));
+        request.GetUnixSocketPath(),
+        request.GetHeaders());
 
+    auto [sessionInfo, error] = Executor->WaitFor(sessionFuture);
     if (HasError(error)) {
         return error;
     }
 
-    auto listenerIt = EndpointListeners.find(oldReq.GetIpcType());
+    auto listenerIt = EndpointListeners.find(request.GetIpcType());
     STORAGE_VERIFY(
         listenerIt != EndpointListeners.end(),
         TWellKnownEntityTypes::ENDPOINT,
-        socketPath);
+        request.GetUnixSocketPath());
 
     auto& listener = listenerIt->second;
 
     auto alterFuture = listener->AlterEndpoint(
-        oldReq,
+        request,
         sessionInfo.Volume,
         sessionInfo.Session);
 
     return Executor->WaitFor(alterFuture);
 }
 
+NProto::TError TEndpointManager::RecreateSocket(
```

Comment:
A bit of confusing to call this `RecreateSocket` where there is nothing socket related in the function. It's better to put a comment explaining that restarting listener endpoint should recreate the socket and it would be good to recheck that socket exists and if it doesn't exist fail the function

Reply:
Yes, I agree. 
So I added a comment, added a recheck that socket exists and also renamed the method.